python/lsst/pipe/base/pipeline.py

```diff
@@ -111,6 +112,18 @@ def __str__(self):
         rep += ")"
         return rep
 
+    def __eq__(self, other: TaskDef) -> bool:
+        return self.taskName == other.taskName and\
+            self.config == other.config and\
+            self.taskClass == other.taskClass and\
+            self.label == other.label
+
+    def __hash__(self):
+        with io.StringIO() as sio:
+            self.config.saveToStream(sio)
+            configHash = hash(sio.getvalue())
```

Comment:
I'm not very comfortable with this, I'm not convinced that formatting pex_config will always produce the same output for two identical configs. Can you just ignore config and not include it in a hash?

Reply:
No, as it is the only thing that can disambiguate a TaskDef created in the same pipeline with the same input repos, only differing by a config option. This is useful for comparing one QuantumGraph to another.

Reply:
The hash is not required to be unique, the requirement is that objects that compare as equal have the same hash, but does not say that objects that are different must have different hash (and hash in general is not unique). You can drop config from hash and still keep it in `__eq__` (comparing configs is well defined). And in the same pipeline tasks have to have different labels so hash that includes labels should already be OK.

Reply:
But the same pipeline might be run two ways, I don't see anything that says the TaskDefs should have the context of anything else other than themselves. I have verified creating a new config and turning it to a string over 2000 times leads to the same serialization. This may not be true from one version of the stack to another, but that is possible with any stack version changes.

Reply:
Shouldn't we be letting `Config` define how it handles equality and `__hash__`?

Reply:
I'm not sure `Config` wants to be hashable. It is a mutable structure unless you freeze it and in general making complicated dynamic data structures hashable is not worth trouble in my opinion. In this particular case we don't even need its hash, other attributes of `TaskDef` are sufficient for building reasonable hash.

Reply:
Any you make a good point, I was so focused on the tree I was not seeing the forest (thinking too closely of equals and not hash interplay). Thanks for pointing out the data model again, equal things must hash the same, but the converse of hash must be everything in equals is not necessary (if you are fine with a bit higher collision rate). I'll update this now.